rollingpin/elasticsearch.py

```diff
@@ -62,56 +62,131 @@ def index_doc(self, doc):
+        timestamp_in_milliseconds = int(time.time()) * 1000
+
+        # We aren't aware of the commit ID that is being deployed at initial
+        # deploy time.  Once the build sync has happened, we have this
+        # information, so we go back and update the original document with a
+        # list of the sync targets (i.e. commit IDs being deployed).
+        sync = ["%s@%s" % (k, v.get('token', '')[:7])
+                for k, v in sync_info.iteritems()]
+        return {
+            'timestamps': {
+                'build': {
+                    'sync': timestamp_in_milliseconds,
+                },
+            },
+            'sync_targets': ', '.join(sync),
+        }
+
     def deploy_start_doc(self):
         timestamp_in_milliseconds = int(time.time()) * 1000
         return {
             'id': self.deploy_name,
+
+            # Once we change the mapping and reindex the data we can remove
+            # this redundant timestamp field.
             'timestamp': timestamp_in_milliseconds,
+
+            'timestamps': {
+                'deploy': {
+                    'start': timestamp_in_milliseconds,
+                },
+            },
             'components': self.components,
             'deployer': getpass.getuser(),
             'command': self.command_line,
             'hosts': self.hosts,
             'host_count': len(self.hosts),
-            'event_type': 'deploy.begin',
         }
 
     def deploy_abort_doc(self, reason):
         timestamp_in_milliseconds = int(time.time()) * 1000
         return {
-            'id': self.deploy_name,
-            'timestamp': timestamp_in_milliseconds,
-            'reason': reason,
-            'components': self.components,
-            'event_type': 'deploy.abort',
+            'timestamps': {
+                'deploy': {
+                    'abort': timestamp_in_milliseconds,
+                },
+            },
+            'abort_reason': reason,
         }
 
     def deploy_end_doc(self):
         timestamp_in_milliseconds = int(time.time()) * 1000
         return {
-            'id': self.deploy_name,
-            'timestamp': timestamp_in_milliseconds,
-            'components': self.components,
-            'event_type': 'deploy.end',
+            'timestamps': {
+                'deploy': {
+                    'end': timestamp_in_milliseconds,
+                },
+            },
         }
 
     @inlineCallbacks
+    def on_build_sync(self, sync_info):
+        if self.deploy_annotation_id:
+            yield self.update_doc(self.deploy_annotation_id,
+                                  self.build_sync_doc(sync_info))
+
+    @inlineCallbacks
     def on_deploy_start(self):
-        yield self.index_doc(self.deploy_start_doc())
+
+        # Store the initial document's ID so we can update it with additional
+        # metadata later in the deploy process.
+        def store_deploy_annotation_id(body):
+            if response.code != 201:
```

Comment:
why is this being checked here where we're reading the body rather than after L167 where we've already read the headers but haven't started the body yet?